File: leetcode/0125.cpp
#include "common.hpp"

namespace leetcode {
    using namespace utilities;
    using std::string;

    bool isPalindromeHelper(string &s, int lo, int hi) {
        while (lo < hi) {
            if (!std::isalnum(s[lo])) lo++;
            else if (!std::isalnum(s[hi])) hi--;
            else if (tolower(s[lo++]) != tolower(s[hi--])) return false;
        }
        return true;
    }

    bool isPalindrome(string s) {
        return isPalindromeHelper(s, 0, s.size() - 1);
    }
}

int main() {
    using namespace leetcode;
    string input = "A man, a plan, a canal: Panama";
    bool output = isPalindrome(input);
    print_inputoutput(input, to_string(output));
}

File: .archived/gocode/h01/0125.go
package h01

func isPalindrome(s string) bool {
	lo, hi := 0, len(s)-1
	for lo < hi {
		if !isAlnum(s[lo]) {
			lo++
		} else if !isAlnum(s[hi]) {
			hi--
		} else if toLower(s[lo]) == toLower(s[hi]) {
			lo++
			hi--
		} else {
			return false
		}
	}
	return true
}

func isAlnum(b byte) bool {
	return (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || (b >= '0' && b <= '9')
	// return unicode.IsLetter(r) || unicode.IsDigit(r)
}

func toLower(b byte) byte {
	if b >= 'A' && b <= 'Z' {
		return b + 32
	}
	return b
}
